pkg/phlaredb/block_querier.go

```diff
@@ -949,6 +987,67 @@ func (b *singleBlockQuerier) SelectMatchingProfiles(ctx context.Context, params
 	return iter.NewMergeIterator(maxBlockProfile, false, iters...), nil
 }
 
+func (b *singleBlockQuerier) LabelNames(ctx context.Context, params *typesv1.LabelNamesRequest) ([]string, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "LabelNames Block")
+	defer sp.Finish()
+
+	err := b.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	selectors, err := parseSelectors(params.Matchers)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := b.index.LabelNames()
+	if err != nil {
+		return nil, err
+	}
+
+	if selectors.matchesAll() {
+		return names, nil
+	}
+
+	// Clear the names slice. We need to run the matchers to decide which
+	// names to keep. We want to keep this slice capacity, as there will
+	// never be more names than what index.LabelNames() returns.
+	names = names[:0]
+
+	for _, matchers := range selectors {
+		iter, err := PostingsForMatchers(b.index, nil, matchers...)
+		if err != nil {
+			return nil, err
+		}
+
+		for iter.Next() {
+			labelNames, err := b.index.LabelNamesFor(iter.At())
+			if err != nil {
+				if err == storage.ErrNotFound {
+					continue
+				}
+				return nil, err
+			}
+
+			// Select all the label names which we don't already have.
+			for _, name := range labelNames {
+				index := sort.SearchStrings(names, name)
```

Comment:
I think you could just use a map here to keep unique names.

Reply:
Actually I think you need to keep only names that intersect with both matchers. Something is weird here.

Reply:
New logic is in https://github.com/grafana/pyroscope/pull/2498/commits/4e3295c94f3028d4ac0155ccfaee9915f6efb2bb.